### ISSUE TITLE
data race in get_cpu_features()

### ISSUE THREAD

**Issue by arvidn** [BLAKE3-team/BLAKE3#220]
according to https://github.com/BLAKE3-team/BLAKE3/issues/99 blake3 is supposed to be safe to use in multi-threaded environments.

However, the [get_cpu_features](https://github.com/BLAKE3-team/BLAKE3/blob/master/c/blake3_dispatch.c#L85) implementation reads from and writes to the global variable `g_cpu_features` *without* synchronization. This triggers an error in thread sanitizer.

This function is called from `blake3_compress_in_place()`, `blake3_compress_xof()`, `blake3_hash_many()` and `blake3_simd_degree()`.

There are a few ways to solve this:

1. document that these functions are not safe to call from multiple different threads
2. make `g_cpu_features` thread local (with the `_Thread_local` storage class specifier)
3. use `call_once()` to ensure `g_cpu_features` is initialized exactly once, and synchronize with all other threads

It seems (3) is most consistent with the intention of the code. Here's a proposed patch to do this:

```
--- a/c/blake3_dispatch.c
+++ b/c/blake3_dispatch.c
@@ -1,6 +1,7 @@
 #include <stdbool.h>
 #include <stddef.h>
 #include <stdint.h>
+#include <threads.h>
 
 #include "blake3_impl.h"
 
@@ -71,20 +72,25 @@ enum cpu_feature {
   UNDEFINED = 1 << 30
 };
 
+static once_flag g_cpu_features_initialized = ONCE_FLAG_INIT;
+
 #if !defined(BLAKE3_TESTING)
 static /* Allow the variable to be controlled manually for testing */
 #endif
     _Atomic enum cpu_feature g_cpu_features = UNDEFINED;
 
+static void set_cpu_features_impl();
+
 #if !defined(BLAKE3_TESTING)
 static
 #endif
     enum cpu_feature
     get_cpu_features() {
+  call_once(&g_cpu_features_initialized, set_cpu_features_impl);
+  return g_cpu_features;
+}
 
-  if (g_cpu_features != UNDEFINED) {
-    return g_cpu_features;
-  } else {
+static void set_cpu_features_impl() {
 #if defined(IS_X86)
     uint32_t regs[4] = {0};
     uint32_t *eax = &regs[0], *ebx = &regs[1], *ecx = &regs[2], *edx = &regs[3];
@@ -123,13 +129,11 @@ static
       }
     }
     g_cpu_features = features;
-    return features;
 #else
     /* How to detect NEON? */
-    return 0;
+    g_cpu_features = 0;
 #endif
   }
-}
 
 void blake3_compress_in_place(uint32_t cv[8],
                               const uint8_t block[BLAKE3_BLOCK_LEN],
```

**Comment by oconnor663** [comment]
Could it be simpler to access the flag with `atomic_store_explicit`/`atomic_load_explicit` and `memory_order_relaxed`? It's no problem if multiple threads repeat the same work in some unlikely case. But are the portability issues with old compilers the same as with `call_once`? @sneves what do you think?

Also, dynamic feature detection is going to remain x86-only for the foreseeable future, so it might make sense to `#ifdef` this code more tightly. We tend to get unused function warnings on non-x86 platforms, and it would be nice to silence those.

A middle ground here could be to use atomics if those are available at compile time, and otherwise to fall back to sketchy non-atomics. If we're only doing this for x86, and also only for pre-C11 compilers, maybe it's "fine"?

**Comment by sneves** [comment]
Simply accessing the flag with atomics won't eliminate the data race. On x86 those loads and stores are already atomic for most intents and purposes.

If you really must avoid the double write, you could use a single atomic compare-exchange replacing the `g_cpu_features = features;` write. This will ensure that the value will only be written once.

Using these thread(-local) C11 features will rule out at least MSVC as a supported compiler. Given that this data race is harmless, another option would be to add the attribute `__attribute__((no_sanitize("thread")))`, conditional on `__SANITIZE_THREAD__`, to the function.

**Comment by oconnor663** [comment]
> Simply accessing the flag with atomics won't eliminate the data race.

My understanding was that reading and writing with atomics was the textbook way to eliminate a data race. Are there other factors I'm missing? I get that atomic loads and stores won't prevent two threads from reading `UNDEFINED` at the same time and then duplicating work, but that seems harmless.

> On x86 those loads and stores are already atomic for most intents and purposes.

I think the soundness issues that @arvidn is referring to are about constraining both the hardware and the compiler. So even if x86 promises to implement ordinary loads and stores in a thread-safe way, in theory GCC could still swoop in and sabotage everything? Maybe reuse the same memory location for temporaries, in some outrageous-but-hey-technically-correct-in-a-single-threaded-context sort of way? But using atomic loads and stores would mean GCC was forbidden by the standard to e.g. invent writes to the shared value.

Anyway all my understanding of this stuff comes from [a YouTube video](https://youtu.be/A8eCGOqgvH4), so I could easily have all this backwards.

**Comment by arvidn** [comment]
my understanding is that using relaxed atomics to read and write would solve the problem. it could lead to computing the value multiple times, which seems harmless. There is no other state that needs to be synchronized in addition to the global variable itself, so relaxed should be sufficient.

I don't think it's sufficient to rely on x86 memory order guarantees, compilers will also optimize assuming the source code has no data races.

After posting this and actually testing my fix on more compilers, I realized that C11 not only rules out msvc, but also MacOS (which supports some of C11, but not the threading functions).

**Comment by oconnor663** [comment]
I tried a simple relaxed-atomics-based implementation on [my `atomic_features` branch](https://github.com/BLAKE3-team/BLAKE3/tree/atomic_features), specifically [this commit](https://github.com/BLAKE3-team/BLAKE3/commit/242ea2b2c887e96f3514e89800626efbc26bfd67). But predictably, that version didn't compile under MSVC, which lacks the `stdatomic.h` header.

It seems like we could potentially work around this by adding a bunch more `#ifdef`s for MSVC, to replace `atomic_store_explicit` with `InterlockedExchangeNoFence` or similar. However, [according to this thread](https://stackoverflow.com/questions/42660091/atomic-load-in-c-with-msvc) MSVC provides no atomic load (?!). That thread links to [Microsoft docs](https://docs.microsoft.com/en-us/windows/win32/sync/interlocked-variable-access) which says, "Simple reads and writes to properly-aligned 32-bit variables are atomic operations." So I guess we're supposed to use ordinary loads and stores as relaxed atomics on that platform. I've added that approach in [this commit](https://github.com/BLAKE3-team/BLAKE3/commit/8b98a264e80b1f93e258605c248612a04ef9a6e7), which passes CI.

On the other hand, if the latest version of MSVC requires compatibility hacks, it's inevitable that older and more obscure compilers are going to require a bunch more hacks. Maybe we could work around this with `__has_include (<stdatomic.h>)`? (Indeed, that's a [suggested example](https://gcc.gnu.org/onlinedocs/cpp/_005f_005fhas_005finclude.html) in GCC's docs.) So platfoms with the new standard stuff gets safe atomics and everyone else gets...silently degraded correctness?

Honestly, I just don't want to ship any of this right now. At best, we'll make the code substantially worse in exchange for an incomplete fix for a bug that's only a bug in theory. But at the same time we'll probably break a couple of obscure targets for who knows what unpredictable reason. It's not worth it. Maybe in another 5 or 10 years the C ecosystem will catch up with the rest of the world, and there will be a correct, widely-supported way to clean this up?

**Comment by odiferousmint** [comment]
Using commit 48c4621edc5be916e47581c3f68dd259a89dbf19 I am already getting the following error when I run `make -f Makefile.testing`:

```
gcc -O3 -Wall -Wextra -std=c11 -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -fPIE -fvisibility=hidden -Wa,--noexecstack blake3.c blake3_dispatch.c blake3_portable.c main.c blake3_sse2.o blake3_sse41.o blake3_avx2.o blake3_avx512.o -o blake3 -pie -Wl,-z,relro,-z,now
/bin/ld: /tmp/ccfZn1c2.o: in function `main':
main.c:(.text.startup+0x2a2): undefined reference to `get_cpu_features'
/bin/ld: main.c:(.text.startup+0x2f5): undefined reference to `g_cpu_features'
collect2: error: ld returned 1 exit status
make: *** [Makefile.testing:50: all] Error 1
```

Is this a bug or am I "holding it wrong"? :)

**Comment by oconnor663** [comment]
@odiferousmint that makefile is really only intended for internal testing, but you can see how we invoke it [here in CI](https://github.com/BLAKE3-team/BLAKE3/blob/master/.github/workflows/ci.yml#L175-L211). In short, the command you'd need to run is `make -f Makefile.testing test` or `make -f Makefile.testing test_asm`. But if you're thinking about making changes to the C code, and you want to run tests while you work, I might recommend `cd c/blake3_c_rust_bindings && cargo test`. That's more likely to give you meaningful feedback when something specific breaks.

**Comment by odiferousmint** [comment]
> @odiferousmint that makefile is really only intended for internal testing, but you can see how we invoke it [here in CI](https://github.com/BLAKE3-team/BLAKE3/blob/master/.github/workflows/ci.yml#L175-L211). In short, the command you'd need to run is `make -f Makefile.testing test` or `make -f Makefile.testing test_asm`. But if you're thinking about making changes to the C code, and you want to run tests while you work, I might recommend `cd c/blake3_c_rust_bindings && cargo test`. That's more likely to give you meaningful feedback when something specific breaks.

Ah, I did not expect to have to run `cargo test` when making changes to the C code, thank you!

**Comment by oconnor663** [comment]
Yeah that might be a quirky personal preference of mine as a Rust fanatic, but I like doing testing and benchmarking of C code via Rust bindings. One of the benefits in this case is that it's easy to test that the Rust and C implementations are producing identical output. It's also easier to just copy-paste tests from the Rust version. :sweat_smile:

**Comment by odiferousmint** [comment]
> Yeah that might be a quirky personal preference of mine as a Rust fanatic, but I like doing testing and benchmarking of C code via Rust bindings. One of the benefits in this case is that it's easy to test that the Rust and C implementations are producing identical output. It's also easier to just copy-paste tests from the Rust version. sweat_smile

Fair enough. I am not sure if BLKAKE3 have always had a C reference implementation. I might misremember here and it may have been another project, but it lacked C reference implementation so I moved on.

The reason being is that C is way easier for me to read and understand; it is almost like pseudo-code to me. I wanted to port something to a language, and I did not have C code to look at, only Rust. I have difficulties with Rust in this regard. :/

**Comment by oconnor663** [comment]
The C implementation is a line-by-line port of the _production_ Rust implementation, and that makes both of them big and complicated (e.g. supporting several SIMD implementations). The simple reference implementation has always been [`reference_impl.rs`](https://github.com/BLAKE3-team/BLAKE3/blob/master/reference_impl/reference_impl.rs). I've [ported that reference to Python here](https://github.com/oconnor663/pure_python_blake3), but I've never ported it to C. Maybe I should go ahead and do that?

**Comment by odiferousmint** [comment]
> The C implementation is a line-by-line port of the _production_ Rust implementation, and that makes both of them big and complicated (e.g. supporting several SIMD implementations). The simple reference implementation has always been [`reference_impl.rs`](https://github.com/BLAKE3-team/BLAKE3/blob/master/reference_impl/reference_impl.rs). I've [ported that reference to Python here](https://github.com/oconnor663/pure_python_blake3), but I've never ported it to C. Maybe I should go ahead and do that?

That would be much appreciated! It would help me (and probably others) a lot!

**Comment by oconnor663** [comment]
@odiferousmint here you go! https://github.com/oconnor663/blake3_reference_impl_c

Let's go ahead and move any further conversation to new issues, to avoid taking this one too far off topic :)

**Comment by mratsim** [comment]
How does feature detection work on Alder Lake with it's heterogenous support, say AVX512? It might to be thread-local.

**Comment by oconnor663** [comment]
@mratsim this is an issue for a specific bug, so please move general questions to new issues or to email. I let myself get off track above, but I don't want to overwhelm the inboxes of folks subscribed to this bug.

**Comment by mratsim** [comment]
My comment is giving more weight to option 2 in OP initial proposal. It is not offtopic.

**Comment by sneves** [comment]
Alder Lake does not enable AVX-512. Even in the case where BIOS had the option to enable it, this was only possible in models where every core had AVX-512 (and it appears now [completely removed](https://www.tomshardware.com/news/intel-nukes-alder-lake-avx-512-now-fuses-it-off-in-silicon)). Every other relevant feature is present on both performance and efficiency cores.

What happens with dynamic CPU feature detection on heterogeneous systems is a good question, but it does not really apply to Alder Lake.

**Comment by p7r0x7** [comment]
For awhile you could disable the efficiency cores in certain BIOSes to access AVX512 that is physically present on the performance cores. I'm not sure whether that's still possible though.
